### ISSUE TITLE
Log likelihood computation in the code

### ISSUE THREAD

**Issue by deepakdalakoti** [cics-nd/rans-uncertainty#3]
Hi,

I have two questions about the way the log-likelihood is computed in the code.

1) For mini-batch training, the log-likelihood is multiplied by a factor of N/B, where N is the total number of training samples and B is the number of samples in the mini-batch. What is the rationale behind this? I know this follows from Ref. [1] but wouldn't this erroneously give more weight to the likelihood function as we decrease the batch size since the prior probability does not change.  This would then be akin to maximizing the likelihood rather than the full aposteriori distribution. What am I missing here?

2) The code is used to predict the Reynolds-stress anisotropy tensor which has 6 unique components. Assuming each component is iid Normal distributed, the joint distribution would be something like:

> P(**b**| **w, beta**) ~ N(**b**|**w, beta**) ~ N(b1|**w, beta**)*N(b2|**w, beta**).....*N(b6|**w, beta**)

where b1,..,b6 are unique components of **b**.

This is the distribution for one sample. For N samples we need to do a similar product N times. 

Now if we look at the code, the likelihood computation reads

```
log_likelihood = len(self.trainingLoader.dataset) / output.size(0) \
                                * (-0.5 * self.models[index].log_beta.exp()
                                * (target - output).pow(2).sum()
                                + 0.5 * target.numel()
                                * self.models[index].log_beta)
``` 

This does not take into account that there are 6 components of **b**. A correct likelihood should instead be

```
log_likelihood = len(self.trainingLoader.dataset) / output.size(0) \
                                * (-0.5 * self.models[index].log_beta.exp()
                                * (target - output).pow(2).sum()
                                + 0.5 * (target.numel()+6)
                                * self.models[index].log_beta)
``` 
where we need to add 6 to the target.numel() term. This constant could have significant effect especially for small mini-batch size. 

[1] Y. Zhu, N. Zabaras. Bayesian deep convolutional encoder–decoder networks for surrogate modeling and uncertainty quantification, Journal of Computational Physics, 2018.

Regards
Deepak

**Comment by NickGeneva** [comment]
Hi,

Thanks for the interest in the code, hopefully I can clear up your questions.

1. From my understanding, the intent of using this scaling is to approximate the likelihood of the entire data-set through a single mini-batch. After all, a mini-batch is just an approximation to a full sized batch (do back-prop using all the data at once) we use due to potential computational constraints and mini-batches work better with gradient descent methods. So for the log-likelihood we pretend each mini-batch has an approximate representation of the full sized data set and scale it up to the same magnitude, hence the N/B.
Will this work well for really small batch sizes, I would guess not really. However, I think the prior would be fine since it was designed for the entire data-set (i.e. the log-likelihood scaled by N/B for a mini-batch). The trouble would be that your basically assuming the entire training data-set being modeled from just a few samples which isn't very accurate. You would have to pray to the gradient descent gods to be nice to you.

2. Yes, you are correct, however I believe the `.numel()` function should take care of this automatically. Since this returns the total number of elements in the PyTorch tensor this should provide the correct value for that factor in the Gaussian.

Hopefully this answers your questions. Let me know otherwise.
